src/step.rs

```diff
@@ -96,57 +100,73 @@ where
     F: Fn(&InterpCx) -> ShouldContinue,
 {
     let mut message = None;
-    loop {
-        if pcx.ecx.stack().len() <= 1 && is_ret(&pcx.ecx) {
-            break;
-        }
-        match pcx.ecx.step() {
-            Ok(true) => {
-                *pcx.step_count += 1;
-                crate::watch::step_callback(pcx);
-
-                if let Some(frame) = pcx.ecx.stack().last() {
-                    if let Some(block) = frame.block {
-                        let blck = &frame.body.basic_blocks()[block];
-                        if frame.stmt != blck.statements.len()
-                            && crate::should_hide_stmt(&blck.statements[frame.stmt])
-                            && !pcx.config.bptree.is_at_breakpoint(&pcx.ecx)
-                        {
-                            continue;
-                        }
-                    } else {
-                        // Unwinding, but no cleanup for this frame
-                        // FIXME make step behaviour configurable
-                    }
+    let ret: InterpResult<'_, _> = (|| {
+        loop {
+            match pcx.ecx.schedule()? {
+                SchedulingAction::ExecuteStep => {}
+                SchedulingAction::ExecuteTimeoutCallback => {
+                    pcx.ecx.run_timeout_callback()?;
+                    continue;
                 }
-                if let ShouldContinue::Stop = continue_while(&pcx.ecx) {
-                    break;
+                SchedulingAction::ExecuteDtors => {
+                    // This will either enable the thread again (so we go back
+                    // to `ExecuteStep`), or determine that this thread is done
+                    // for good.
+                    pcx.ecx.schedule_next_tls_dtor_for_active_thread()?;
+                    continue;
                 }
-                if pcx.config.bptree.is_at_breakpoint(&pcx.ecx) {
+                SchedulingAction::Stop => {
+                    message = Some("interpretation finished".to_string());
                     break;
                 }
             }
-            Ok(false) => {
-                message = Some("interpretation finished".to_string());
+            let info = pcx.ecx.preprocess_diagnostics();
+            assert!(pcx.ecx.step()?, "a terminated thread was scheduled for execution");
```

Comment:
Whenever possible please report any unexpected things to the user and continue. It shouldn't just crash on unexpected events, as that would make it much less useful for debugging miri errors.